cloudprovider/aws/aws_provider.go

```diff
@@ -37,75 +29,45 @@ type metadataClientImpl struct {
 	client *ec2metadata.EC2Metadata
 }
 
+func init() {
+	cloudprovider.AddCloudProvider(awsTag, &Provider{
+		expireTime: time.Minute * 3,
+		interval:   time.Microsecond * 500,
+	})
+}
+
 func (m metadataClientImpl) getInstanceIdentityDocument() (ec2metadata.EC2InstanceIdentityDocument, error) {
 	return m.client.GetInstanceIdentityDocument()
 }
 
-func NewProvider() Provider {
-	prod := Provider{}
+func (p *Provider) Init() error {
 	s, err := session.NewSession()
 	if err != nil {
 		logrus.Error(err)
-		return Provider{}
+		return err
```

Comment:
If an error is returned by the function, you don't need to put the same error in the log inside the function. The caller should deal with that.